packages/datadog-plugin-fs/src/index.js

```diff
@@ -0,0 +1,497 @@
+          return read.call(this, (err, dirent) => {
+            done(err)
+            cb(err, dirent)
+          })
+        } else {
+          const p = read.call(this)
+          if (p && typeof p.then === 'function') {
+            return p.then(r => {
+              done()
+              return r
+            }, e => {
+              done(e)
+              throw e
+            })
+          } else {
+            done()
+            return p
+          }
+        }
+      })
+    }
+  }
+}
+
+function createWrapDirClose (config, tracer, sync) {
+  const name = sync ? 'fs.dir.closesync' : 'fs.dir.close'
+  return function wrapDirClose (close) {
+    return function wrappedDirClose (cb) {
+      const tags = makeFSTags(this.path, null, null, config, tracer)
+      return tracer.trace(name, { tags }, (span, done) => {
+        if (cb) {
+          return close.call(this, (err, dirent) => {
+            done(err)
+            cb(err, dirent)
+          })
+        } else {
+          const p = close.call(this)
+          if (p && typeof p.then === 'function') {
+            return p.then(r => {
+              done()
+              return r
+            }, e => {
+              done(e)
+              throw e
+            })
+          } else {
+            done()
+            return p
+          }
+        }
+      })
+    }
+  }
+}
+
+let kDirReadPromisified
+let kDirClosePromisified
+
+function createWrapDirAsyncIterator (config, tracer, instrumenter) {
+  return function wrapDirAsyncIterator (asyncIterator) {
+    return function wrappedDirAsyncIterator () {
+      if (!kDirReadPromisified) {
+        const keys = Reflect.ownKeys(this)
+        for (const key of keys) {
+          if (kDirReadPromisified && kDirClosePromisified) break
+          if (typeof key !== 'symbol') continue
+          if (!kDirReadPromisified && getSymbolName(key).includes('kDirReadPromisified')) {
+            kDirReadPromisified = key
+          }
+          if (!kDirClosePromisified && getSymbolName(key).includes('kDirClosePromisified')) {
+            kDirClosePromisified = key
+          }
+        }
+      }
+      instrumenter.wrap(this, kDirReadPromisified, createWrapKDirRead(config, tracer))
+      instrumenter.wrap(this, kDirClosePromisified, createWrapKDirClose(config, tracer, instrumenter))
+      return asyncIterator.call(this)
+    }
+  }
+}
+
+function createWrapKDirRead (config, tracer) {
+  return createWrapDirRead(config, tracer)
+}
+
+function createWrapKDirClose (config, tracer, instrumenter) {
+  return function wrapKDirClose (kDirClose) {
+    return function wrappedKDirClose () {
+      const tags = makeFSTags(this.path, null, null, config, tracer)
+      return tracer.trace('fs.dir.close', { tags }, (span) => {
+        const p = kDirClose.call(this)
+        return p.then(r => {
+          instrumenter.unwrap(this, kDirReadPromisified)
+          instrumenter.unwrap(this, kDirClosePromisified)
+          return r
+        }, e => {
```

Comment:
Sort of. The second one is the `catch`. The idea here is we want to unwrap regardless of resolve or reject. That being said, all of this can go away if we don't need to unwrap.

Reply:
Since the code is identical, it could be a single function passed twice instead of 2 identical functions. `p` should also be returned directly and not the result of `p.then` since it would alter the return value.

Reply:
The functions *aren't* identical. One of them returns, the other throws in order to maintain the rejection status.

If the original promise should be returned, then it doesn't matter though, so I'll go ahead and change that.